// @hediet/node-reload/dist/hotDecorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const HotReloadService_1 = require("./HotReloadService");
class FunctionStore {
    constructor() {
        this.map = new Map();
    }
    format(module, className, methodName) {
        return JSON.stringify({ mod: module.filename, className, methodName });
    }
    setFunc(module, className, methodName, fn) {
        const key = this.format(module, className, methodName);
        this.map.set(key, fn);
    }
    getFunc(module, className, methodName) {
        const key = this.format(module, className, methodName);
        return this.map.get(key);
    }
}
FunctionStore.instance = new FunctionStore();
function hotMethod(module, options = { restartOnReload: false }) {
    return function (target, propertyKey, descriptor) {
        const instance = HotReloadService_1.HotReloadService.instance;
        if (!instance) {
            return;
        }
        const func = target[propertyKey];
        const className = target.constructor.name;
        FunctionStore.instance.setFunc(module, className, propertyKey, func);
        target[propertyKey] = function (...args) {
            let result;
            while (true) {
                const newFunc = FunctionStore.instance.getFunc(module, className, propertyKey);
                result = newFunc.apply(this, args);
                if (options.restartOnReload &&
                    instance.handleFileMightHaveChanged(module.filename)) {
                    continue;
                }
                break;
            }
            return result;
        };
    };
}
exports.hotMethod = hotMethod;
